nixos/modules/services/development/jupyter.nix

```diff
@@ -0,0 +1,236 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.jupyter;
+  jupyterHome = pkgs.stdenv.mkDerivation rec {
+    name = "jupyter-kernels";
+
+    src = "/dev/null";
+    unpackCmd ="mkdir jupyter_kernels";
+    installPhase =
+      ''
+        mkdir kernels
+        ${concatStringsSep "\n" (mapAttrsToList (kernelName: kernel:
+          let
+            config = builtins.toJSON {
+              argv = ([ "${kernel.env}/bin/${kernel.executable}" ] ++ kernel.argv);
+              display_name = kernel.name;
+              language = kernel.language;
+            };
+          in ''
+            mkdir kernels/${kernelName};
+            echo '${config}' > kernels/${kernelName}/kernel.json;
+          '') cfg.kernels)}
+
+        mkdir $out
+        cp -r kernels $out
+      '';
+  };
+
+  notebookConfig = pkgs.writeText "jupyter_config.py" ''
+    c.NotebookApp.password = u'${cfg.password}'
+
+    ${cfg.notebookConfig}
+  '';
+
+in {
+
+  options.services.jupyter = {
+    enable = mkEnableOption "Jupyter development server";
+
+    ip = mkOption {
+      type = types.str;
+      default = "localhost";
+      description = ''
+        IP address Jupyter will be listening on.
+      '';
+    };
+
+    port = mkOption {
+      type = types.int;
+      default = 8888;
+      description = ''
+        Port number Jupyter will be listening on.
+      '';
+    };
+
+    notebookDir = mkOption {
+      type = types.str;
+      default = "~/";
+      description = ''
+        Root directory for notebooks.
+      '';
+    };
+
+    user = mkOption {
+      type = types.str;
+      default = "jupyter";
+      description = ''
+        Name of the user used to run the jupyter service.
+        For security reason, jupyter should really not be run as root.
+        If not set (jupyter), the service will create a jupyter user with appropriate settings.
+      '';
+      example = "aborsu";
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = "jupyter";
+      description = ''
+        Name of the group used to run the jupyter service.
+        Use this if you want to create a group of users that are able to view the notebook directory's content.
+      '';
+      example = "users";
+    };
+
+    password = mkOption {
```

Comment:
I think it is better to keep this out of the store. Maybe you can point to a path and read that during runtime?

Reply:
I made a compromise on that, that I think should be agreeable to all.